pkg/gatewayserver/redis/registry.go

```diff
@@ -27,28 +31,127 @@ type GatewayConnectionStatsRegistry struct {
 	Redis *ttnredis.Client
 }
 
-func (r *GatewayConnectionStatsRegistry) key(uid string) string {
-	return r.Redis.Key("uid", uid)
+var (
+	down        = "down"
+	up          = "up"
+	status      = "status"
+	errNotFound = errors.DefineNotFound("stats_not_found", "Gateway Stats not found")
+)
+
+func (r *GatewayConnectionStatsRegistry) key(which string, uid string) string {
+	return r.Redis.Key(which, "uid", uid)
 }
 
 // Set sets or clears the connection stats for a gateway.
-func (r *GatewayConnectionStatsRegistry) Set(ctx context.Context, ids ttnpb.GatewayIdentifiers, stats *ttnpb.GatewayConnectionStats) error {
+func (r *GatewayConnectionStatsRegistry) Set(ctx context.Context, ids ttnpb.GatewayIdentifiers, stats *ttnpb.GatewayConnectionStats, update io.Traffic) error {
 	uid := unique.ID(ctx, ids)
+
+	defer trace.StartRegion(ctx, "set gateway connection stats").End()
+
+	var err error
 	if stats == nil {
-		return r.Redis.Del(r.key(uid)).Err()
+		_, err = r.Redis.Pipelined(func(p redis.Pipeliner) error {
+			if update.Up {
+				if err := p.Del(r.key(up, uid)).Err(); err != nil {
+					return err
+				}
+			}
+			if update.Down {
+				if err := p.Del(r.key(down, uid)).Err(); err != nil {
+					return err
+				}
+			}
+			if update.Status || (!update.Up && !update.Down) {
+				if err := p.Del(r.key(status, uid)).Err(); err != nil {
+					return err
+				}
+			}
+			return nil
+		})
+	} else {
+		// Update (pipelined for better performance) otherwise
+		_, err = r.Redis.Pipelined(func(p redis.Pipeliner) error {
+			if update.Up {
+				if _, err := ttnredis.SetProto(p, r.key(up, uid), stats, 0); err != nil {
+					return err
+				}
+			}
+			if update.Down {
+				if _, err := ttnredis.SetProto(p, r.key(down, uid), stats, 0); err != nil {
+					return err
+				}
+			}
+			if update.Status || (!update.Up && !update.Down) {
+				if _, err := ttnredis.SetProto(p, r.key(status, uid), stats, 0); err != nil {
+					return err
+				}
+			}
+			return nil
+		})
+
 	}
 
-	_, err := ttnredis.SetProto(r.Redis, r.key(uid), stats, 0)
-	return err
+	if err != nil {
+		return ttnredis.ConvertError(err)
+	}
+	return nil
 }
 
 // Get returns the connection stats for a gateway.
 func (r *GatewayConnectionStatsRegistry) Get(ctx context.Context, ids ttnpb.GatewayIdentifiers) (*ttnpb.GatewayConnectionStats, error) {
 	uid := unique.ID(ctx, ids)
+	result := &ttnpb.GatewayConnectionStats{}
 	stats := &ttnpb.GatewayConnectionStats{}
-	err := ttnredis.GetProto(r.Redis, r.key(uid)).ScanProto(stats)
+
+	retrieved, err := r.Redis.MGet(r.key(up, uid), r.key(down, uid), r.key(status, uid)).Result()
 	if err != nil {
-		stats = nil
+		return nil, ttnredis.ConvertError(err)
 	}
-	return stats, err
+
+	if retrieved[0] == nil && retrieved[1] == nil && retrieved[2] == nil {
+		return nil, errNotFound
+	}
+
+	// uplink stats
+	if retrieved[0] != nil {
+		if ttnredis.UnmarshalProto(retrieved[0].(string), stats) == nil {
```

Comment:
Use proper error handling. Also, if this fails, that's not a `errNotFound` but a corrupt store